### PR TITLE
Wizard of Wikipedia generative models

### PR DESCRIPTION
**Pull request by stephenroller**
Generative models from Wizard of Wikipedia paper.

Some issues:
- [x] Generalize how model args were being passed in for TGA.
- [x] Generalize loss computations bit
- [x] Generalize 
- [ ] ~~Bring in hierarchical agent somewhere~~ Wait until we get approval for pretrained ranker release
- [x] Pretrained model in model zoo
- [ ] Integration test of end2end model

Needs #1487 to merge first.

### CODE REVIEW

File: projects/wizard_of_wikipedia/generator/tasks.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        if self.only_checked_knowledge:
+            # useful for test time evaluation, where it's only ever trained on true
+            # knowledge
+            a['knowledge'] = (
+                a['title'] + ' ' + TOKEN_KNOWLEDGE + ' ' + a['checked_sentence']
+            )
+        return a
+
+
+class GoldWizardTeacher(WizardTeacher):
```
Comment (huihuifan) [inline projects/wizard_of_wikipedia/generator/tasks.py]: is it preferred in parlai to have a more very specific teachers, or one general teacher that can take in different parameters like 'only_checked_knowledge' or 'ignorant_dropout'?

File: projects/wizard_of_wikipedia/generator/agents.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+            return list(obs['knowledge_parsed'])
+
+        checked_sentence = '{} {} {}'.format(
+            obs['title'], TOKEN_KNOWLEDGE, obs['checked_sentence']
+        )
+        # grab all the nonempty knowledge
+        obs_know = [k.strip() for k in obs.get('knowledge', '').split('\n')]
+        obs_know = [k for k in obs_know if k]
+
+        # we want the correct knowledge to always be in index 0
```
Comment (huihuifan) [inline projects/wizard_of_wikipedia/generator/agents.py]: I'm a little lost on this part. Why do you force it to be first?

Comment (stephenroller) [reply projects/wizard_of_wikipedia/generator/agents.py]: IIRC later on I use that fact to make loss computation easier or something.

File: projects/wizard_of_wikipedia/generator/tasks.py (hunk 1 above)
Comment (stephenroller) [reply projects/wizard_of_wikipedia/generator/tasks.py]: I’m not sure; CLI parameters are more common. This was just a convenient way for me to do it at the time.

File: parlai/core/torch_generator_agent.py [hunk 3]
```diff
@@ -451,7 +461,9 @@ def train_step(self, batch):
         self.zero_grad()
 
         try:
-            scores, preds, _ = self.model(batch.text_vec, batch.label_vec)
+            scores, preds, *_ = self.model(
+                *self._model_input(batch), ys=batch.label_vec
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: cc @alexholdenmiller

**Comment by stephenroller** [comment]
Seeking reviews

File: parlai/core/torch_generator_agent.py [hunk 4]
```diff
@@ -211,13 +214,12 @@ def forward(self, xs, ys=None, cand_params=None, prev_enc=None, maxlen=None):
...
+        encoder_states = prev_enc if prev_enc is not None else self.encoder(*xs)
 
         if ys is not None:
             # use teacher forcing
             scores, preds = self.decode_forced(encoder_states, ys)
         else:
-            bsz = xs.size(0)
             scores, preds = self.decode_greedy(
                 encoder_states,
                 bsz,
```
Comment (klshuster) [inline parlai/core/torch_generator_agent.py]: does this protect against bsz being None?

File: projects/wizard_of_wikipedia/generator/agents.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+            fields += [obs['title']]
+        if 'checked_sentence' in obs:
+            fields += [TOKEN_KNOWLEDGE, obs['checked_sentence']]
+        if obs['text'] != '':
+            fields += [TOKEN_DIALOG, obs['text']]
+        obs['text'] = ' '.join(fields)
+
+        # now vectorize with the extra knowledge. It'll all get stored in the
+        # text_vec operation, etc
+        self.vectorize(
```
Comment (klshuster) [inline projects/wizard_of_wikipedia/generator/agents.py]: i guess it may not be *necessary* but is it worth setting `obs = self.vectorize(...)` rather than just calling it here? it's not 100% intuitive that vectorize modifies `obs`

File: projects/wizard_of_wikipedia/generator/agents.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        )
+        # grab all the nonempty knowledge
+        obs_know = [k.strip() for k in obs.get('knowledge', '').split('\n')]
+        obs_know = [k for k in obs_know if k]
+
+        # we want the correct knowledge to always be in index 0
+        try:
+            i = obs_know.index(checked_sentence)
+        except ValueError:
+            # uh oh, couldn't find the sentence in the knowledge. This happens for
```
Comment (klshuster) [inline projects/wizard_of_wikipedia/generator/agents.py]: does this really happen still??

Comment (stephenroller) [reply projects/wizard_of_wikipedia/generator/agents.py]: Yes, I think it happens once. That particular sentence didn't have a title associated or something.

File: parlai/core/torch_generator_agent.py (hunk 4 above)
Comment (stephenroller) [reply parlai/core/torch_generator_agent.py]: When is bsz None? greedy_decode needs to be explicitly given the batch size. It uses `batch.text_vec.size(0)`.

Comment (klshuster) [reply parlai/core/torch_generator_agent.py]: I more meant that `bsz` is an optional arg here, and may be `None` - is this ensured anywhere that `bsz` is in fact not `None` when `ys` is? E.g. I found a snippet [here](https://github.com/facebookresearch/ParlAI/blob/a2535496ec454822b83dea7525ceaa4e856dfb22/parlai/core/torch_generator_agent.py#L744) that calls the model with potentially `ys` as `None` and `bsz` as `None`

Comment (stephenroller) [reply parlai/core/torch_generator_agent.py]: Honestly I think that class is nearly entirely broken... But it's a good point. Let's move this into #1469.
